server/src/main/java/timely/store/iterators/TimeSeriesGroupingIterator.java

```diff
@@ -0,0 +1,309 @@
+
+        private static final long serialVersionUID = 1L;
+
+        public TimeSeriesGroup() {
+            super(new TimeSeriesMetricComparator());
+        }
+
+        /**
+         * Iterates over the time series group and returns for each time series
+         * the last key and the filtered value.
+         */
+        @Override
+        public Iterator<Pair<Key, Double>> iterator() {
+
+            final Iterator<Entry<Metric, TimeSeries>> iter = this.entrySet().iterator();
+
+            return new Iterator<Pair<Key, Double>>() {
+
+                @Override
+                public boolean hasNext() {
+                    return iter.hasNext();
+                }
+
+                @Override
+                public Pair<Key, Double> next() {
+                    Entry<Metric, TimeSeries> e = iter.next();
+                    LOG.trace("Removing first entry from series {}", e.getValue());
+                    get(e.getKey()).pollFirstEntry(); // remove first key, will
+                                                      // re-fill later
+                    return new Pair<Key, Double>(e.getValue().lastKey(), e.getValue().getAnswer());
+                }
+
+            };
+        }
+    }
+
+    private static final Logger LOG = LoggerFactory.getLogger(TimeSeriesGroupingIterator.class);
+
+    public static final String FILTER = "sliding.window.filter";
+
+    private final TimeSeriesGroup series = new TimeSeriesGroup();
+    protected Double[] filters = null;
+    private Key topKey = null;
+    private Value topValue = null;
+    private Iterator<Pair<Key, Double>> seriesIterator = null;
+
+    protected Double compute(Map<Key, Double> values) {
+        double result = 0D;
+        int i = 0;
+        for (Entry<Key, Double> e : values.entrySet()) {
+            LOG.trace("compute - key:{}, value: {}", e.getKey(), e.getValue());
+            result += (filters[i] * e.getValue());
+            i++;
+        }
+        LOG.trace("compute - result: {}", result);
+        return result;
+    }
+
+    @Override
+    public Key getTopKey() {
+        return topKey;
+    }
+
+    @Override
+    public Value getTopValue() {
+        return topValue;
+    }
+
+    @Override
+    public boolean hasTop() {
+        LOG.trace("hasTop()");
+        if (seriesIterator.hasNext()) {
+            LOG.trace("seriesIterator - hasNext()");
+            Pair<Key, Double> p = seriesIterator.next();
+            topKey = p.getFirst();
+            topValue = new Value(MetricAdapter.encodeValue(p.getSecond()));
+            return true;
+        } else {
+            try {
+                refillBuffer();
+                if (series.size() == 0) {
+                    return false;
+                } else {
+                    seriesIterator = series.iterator();
+                    Pair<Key, Double> p = seriesIterator.next();
+                    topKey = p.getFirst();
```

Comment:
We could reorganize the code to avoid these 3 lines being repeated:
if (!seriesIterator.hasNext()) {
  refillBuffer()
  reset iterator
}
if (seriesIterator.hasNext()) {
  ...
  return true;
} else {
  return false;
}